expression/builtin_op.go

```diff
@@ -391,6 +407,203 @@ func (b *builtinUnaryOpSig) eval(row []types.Datum) (d types.Datum, err error) {
 	return
 }
 
+type unaryMinusFunctionClass struct {
+	baseFunctionClass
+}
+
+func (b *unaryMinusFunctionClass) typeInfer(argExpr Expression, bf *baseBuiltinFunc) bool {
+	bf.tp.Init(mysql.TypeLonglong)
+	switch argExpr.GetType().Tp {
+	case mysql.TypeString, mysql.TypeVarchar, mysql.TypeVarString, mysql.TypeDouble, mysql.TypeFloat,
+		mysql.TypeDatetime, mysql.TypeDuration, mysql.TypeTimestamp:
+		bf.tp.Tp = mysql.TypeDouble
+	case mysql.TypeNewDecimal:
+		bf.tp.Tp = mysql.TypeNewDecimal
+	}
+	types.SetBinChsClnFlag(bf.tp)
+
+	overflow := false
+	if arg, ok := argExpr.(*Constant); ok {
+		switch arg.Value.Kind() {
+		case types.KindUint64:
+			uval := arg.Value.GetUint64()
+			if uval > uint64(-math.MinInt64) {
+				overflow = true
+				bf.tp.Tp = mysql.TypeNewDecimal
+			}
+		case types.KindInt64:
+			val := arg.Value.GetInt64()
+			if val == math.MinInt64 {
+				overflow = true
+				bf.tp.Tp = mysql.TypeNewDecimal
+			}
+		}
+	}
+	return overflow
+}
+
+func (b *unaryMinusFunctionClass) getFunction(args []Expression, ctx context.Context) (sig builtinFunc, err error) {
+	bf := newBaseBuiltinFunc(args, ctx)
+	argExpr := args[0]
+	overflow := b.typeInfer(argExpr, &bf)
+	if overflow {
+		sig = &builtinUnaryMinusDecimalSig{baseDecimalBuiltinFunc{bf}, true}
+		return sig.setSelf(sig), errors.Trace(b.verifyArgs(args))
+	}
+
+	switch argExpr.GetTypeClass() {
+	case types.ClassInt:
+		sig = &builtinUnaryMinusIntSig{baseIntBuiltinFunc{bf}}
+	case types.ClassDecimal:
+		sig = &builtinUnaryMinusDecimalSig{baseDecimalBuiltinFunc{bf}, false}
+	case types.ClassReal:
+		sig = &builtinUnaryMinusRealSig{baseRealBuiltinFunc{bf}}
+	case types.ClassString:
+		tp := argExpr.GetType().Tp
+		if types.IsTypeTime(tp) {
+			sig = &builtinUnaryMinusDecimalSig{baseDecimalBuiltinFunc{bf}, false}
+		} else if tp == mysql.TypeDuration {
+			sig = &builtinUnaryMinusDecimalSig{baseDecimalBuiltinFunc{bf}, false}
+		} else {
+			sig = &builtinUnaryMinusRealSig{baseRealBuiltinFunc{bf}}
+		}
+	}
+
+	return sig.setSelf(sig), errors.Trace(b.verifyArgs(args))
+}
+
+func unaryMinusDecimal(dec *types.MyDecimal) (to *types.MyDecimal, err error) {
+	to = new(types.MyDecimal)
+	err = types.DecimalSub(new(types.MyDecimal), dec, to)
+	return
+}
+
+type builtinUnaryMinusIntSig struct {
+	baseIntBuiltinFunc
+}
+
+// func (b *builtinUnaryMinusIntSig) eval(row []types.Datum) (d types.Datum, err error) {
```

Comment:
yes, this is not the final commit, cause some test case failed I can't fix, so I push the code to github, I will these code after problem is fixed